### ISSUE TITLE
Problems encountered in implementing few-shot nerf synthesis by hash coding.

### ISSUE THREAD

**Issue by BCL123456-BAL** [FDUEDREF3/FreeNeRF_NeRFStudio#2]
Hey friend. I'm trying to use torch-ngp to realize few-shot nerf new view synthesis, and refer to your code, but the effect is very poor, can you help me?
The grid parameters are as follows：
class NeRFNetwork(NeRFRenderer):
    def __init__(self,
                 encoding="hashgrid",
                 encoding_dir="sphere_harmonics",
                 #encoding_dir="frequency",
                 encoding_bg="hashgrid",   
                 num_layers=2,
                 hidden_dim=64,
                 geo_feat_dim=15,    
                 num_layers_color=3,
                 hidden_dim_color=64,
                 num_layers_bg=2,   
                 hidden_dim_bg=64,   
                 bound=1,
                 **kwargs,
                 ):
self.encoder, self.in_dim = get_encoder(encoding, desired_resolution=2048 * bound)

        sigma_net = []
        for l in range(num_layers):
            if l == 0:
                in_dim = self.in_dim
            else:
                in_dim = hidden_dim
            
            if l == num_layers - 1:
                out_dim = 1 + self.geo_feat_dim # 1 sigma + 15 SH features for color
            else:
                out_dim = hidden_dim
            
            sigma_net.append(nn.Linear(in_dim, out_dim, bias=False))

        self.sigma_net = nn.ModuleList(sigma_net)

        # color network
        self.num_layers_color = num_layers_color        
        self.hidden_dim_color = hidden_dim_color
        self.encoder_dir, self.in_dim_dir = get_encoder(encoding_dir)
        
        color_net =  []
        for l in range(num_layers_color):
            if l == 0:
                in_dim = self.in_dim_dir + self.geo_feat_dim
            else:
                in_dim = hidden_dim_color
            
            if l == num_layers_color - 1:
                out_dim = 3 # 3 rgb  
            else:
                out_dim = hidden_dim_color
            
            color_net.append(nn.Linear(in_dim, out_dim, bias=False))

        self.color_net = nn.ModuleList(color_net)

class GridEncoder(nn.Module):
    def __init__(self, input_dim=3, num_levels=16, level_dim=2, per_level_scale=2, base_resolution=16, log2_hashmap_size=19,
                          desired_resolution=None, gridtype='hash', align_corners=False, interpolation='linear'):

#########Implement regularization here#############
    def forward(self, x, d, global_step=0):
        # x: [N, 3], in [-bound, bound]
        # d: [N, 3], nomalized in [-1, 1]

        # sigma
        x = self.encoder(x, bound=self.bound)

        Total_iters = 30000
        num_views = 9

        if num_views == 3:
            T = int(0.9*Total_iters)
        elif num_views == 6:
            T = int(0.7*Total_iters)
        else:
            T = int(0.2*Total_iters)

        h = x    # h={Tensor(925568,32)}
        X_len = h.shape[-1]
        pos_freq_mask = get_freq_mask(X_len, global_step, T)
        pos_freq_mask = pos_freq_mask.to(h.device)

        h = h * pos_freq_mask.view(1, -1)
        for l in range(self.num_layers):
            h = self.sigma_net[l](h)
            if l != self.num_layers - 1:
                h = F.relu(h, inplace=True)

        #sigma = F.relu(h[..., 0])
        sigma = trunc_exp(h[..., 0])
        geo_feat = h[..., 1:]

        # color
        
        d = self.encoder_dir(d)  #d={Tensor:(2073728,16)}
        D_len = d.shape[-1]
        dir_freq_mask = get_freq_mask(D_len, global_step, T)
        dir_freq_mask = dir_freq_mask.to(d.device)
        d = d * dir_freq_mask.view(1,-1)
        #d[:,int(t/T*D_len)+3:] = 0
        h = torch.cat([d, geo_feat], dim=-1)
        for l in range(self.num_layers_color):
            h = self.color_net[l](h)
            if l != self.num_layers_color - 1:
                h = F.relu(h, inplace=True)
        
        # sigmoid activation for rgb
        color = torch.sigmoid(h)

        return sigma, color

def get_freq_mask(pos_enc_length, current_iter, total_reg_iter):
    if current_iter < total_reg_iter:
        freq_mask=torch.zeros(pos_enc_length)
        ptr = pos_enc_length / 3 * current_iter / total_reg_iter + 1
        ptr = ptr if ptr < pos_enc_length / 3 else pos_enc_length / 3
        int_ptr = int(ptr)
        freq_mask[: int_ptr * 3] = 1.0  # assign the integer part
        freq_mask[int_ptr * 3 : int_ptr * 3 + 3] = (ptr - int_ptr)  # assign the fractional part
        return torch.clip(freq_mask,1e-8, 1-1e-8) # for math stabiltiy
    else:
        return torch.ones(pos_enc_length)

**Comment by BCL123456-BAL** [comment]
i have solved it

**Comment by Mabel0403** [comment]
Hello @BCL123456-BAL  , may I know which version of nerfstudio you are using？I tried a lot of versions of nerfstudio, but each version is not compatible with freenerf.
